File: packages/antcode_core/src/antcode_core/domain/schemas/crawl.py
"""爬取批次相关的 Pydantic 模式定义"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class _UniqueSeedURLs(BaseModel):
    @field_validator("seed_urls", check_fields=False)
    @classmethod
    def reject_duplicate_seed_urls(cls, value: list[str]) -> list[str]:
        if len(value) != len(set(value)):
            raise ValueError("seed_urls 不允许重复")
        return value


# =============================================================================
# 请求模型
# =============================================================================


class CrawlBatchCreateRequest(_UniqueSeedURLs):
    """创建爬取批次请求"""

    project_id: str = Field(..., description="项目公开ID")
    name: str = Field(..., min_length=1, max_length=255, description="批次名称")
    description: str = Field("", max_length=1000, description="批次描述")
    # P1-29 集合大小兜底:1w 个种子 URL 已远超真实业务需要,超过就 422,
    # 避免 10M 条 URL 打爆 asyncpg / Pydantic 反序列化 OOM。
    seed_urls: list[str] = Field(..., min_length=1, max_length=10000, description="种子URL列表(最多 10000)")
    max_depth: int = Field(3, ge=1, le=10, description="最大爬取深度")
    max_pages: int = Field(10000, ge=1, le=1000000, description="最大爬取页面数")
    max_concurrency: int = Field(50, ge=1, le=500, description="最大并发数")
    request_delay: float = Field(0.5, ge=0, le=60, description="请求间隔(秒)")
    timeout: int = Field(30, ge=1, le=300, description="请求超时(秒)")
    max_retries: int = Field(3, ge=0, le=10, description="最大重试次数")


class CrawlBatchTestRequest(_UniqueSeedURLs):
    """测试执行请求"""

    project_id: str = Field(..., description="项目公开ID")
    seed_urls: list[str] = Field(..., min_length=1, max_length=10, description="种子URL列表")
    max_depth: int = Field(2, ge=1, le=3, description="最大爬取深度")
    max_pages: int = Field(10, ge=1, le=100, description="最大爬取页面数")


# =============================================================================
# 响应模型
# =============================================================================


class CrawlBatchResponse(BaseModel):
    """爬取批次响应"""

    id: str = Field(..., description="批次公开ID")
    project_id: str = Field(..., description="项目公开ID")
    name: str = Field(..., description="批次名称")
    description: str = Field("", description="批次描述")
    seed_urls: list[str] = Field(..., description="种子URL列表")
    max_depth: int = Field(..., description="最大爬取深度")
    max_pages: int = Field(..., description="最大爬取页面数")
    max_concurrency: int = Field(..., description="最大并发数")
    request_delay: float = Field(..., description="请求间隔(秒)")
    timeout: int = Field(..., description="请求超时(秒)")
    max_retries: int = Field(..., description="最大重试次数")
    status: str = Field(..., description="批次状态")
    is_test: bool = Field(..., description="是否为测试批次")
    created_at: datetime = Field(..., description="创建时间")
    # S7 stress fix: model 里这俩字段是 nullable datetime，Pydantic 会
    # 拿到 datetime 或 None。之前定义 str 让 running 的 batch 详情端点直接
    # 500，也没被 list 端点覆盖（新建的 batch 全 None 反而通过）。改成
    # datetime | None 让 Pydantic 自动序列化 ISO 字符串。
    started_at: datetime | None = Field(None, description="开始时间")
    completed_at: datetime | None = Field(None, description="完成时间")

    model_config = ConfigDict(from_attributes=True)


class BatchProgressResponse(BaseModel):
    """批次进度响应"""

    batch_id: str = Field(..., description="批次公开ID")
    total_urls: int = Field(0, description="总URL数")
    pending_urls: int = Field(0, description="待处理数")
    completed_urls: int = Field(0, description="已完成数")
    failed_urls: int = Field(0, description="失败数")
    active_workers: int = Field(0, description="活跃Worker数")
    speed_per_minute: float = Field(0.0, description="速度(URLs/分钟)")
    last_updated: str | None = Field(default=None, description="最后更新时间")


class CrawlTestResultResponse(BaseModel):
    """测试执行结果响应"""

    batch_id: str = Field(..., description="测试批次ID")
    success: bool = Field(..., description="是否成功")
    total_pages: int = Field(0, description="爬取页面数")
    success_pages: int = Field(0, description="成功页面数")
    failed_pages: int = Field(0, description="失败页面数")
    sample_data: list[dict] = Field(default_factory=list, description="样本数据")
    errors: list[str] = Field(default_factory=list, description="错误信息列表")
    duration_seconds: float = Field(0.0, description="执行耗时(秒)")


class TestStatusResponse(BaseModel):
    """测试状态响应"""

    batch_id: str = Field("", description="批次ID")
    status: str = Field(..., description="状态(pending/running/completed/failed/not_found)")
    progress: BatchProgressResponse | None = Field(default=None, description="进度信息")
    started_at: str | None = Field(default=None, description="开始时间")
    completed_at: str | None = Field(default=None, description="完成时间")

File: web/antcode-frontend/src/services/crawl.ts
/**
 * Crawl 批次 API 客户端 (R1-P2-28)
 *
 * 后端 18 个批次/指标端点齐全但此前前端 0 引用 —— 用户完全无入口。
 * 这里补上 crawl service 让 UI 页面能查/建/暂停/恢复/取消批次 +
 * 聚合抓取数据 + CSV/JSON 导出。
 */
import { BaseService } from './base'

export interface CrawlBatchSummary {
  // 后端 CrawlBatchResponse.id 承载的是 batch.public_id（对齐 crawl.py:110）；
  // 前端别再自造 batch_id 名字，历史上因此走出 /batches/undefined/... 请求。
  id: string
  project_id: string
  name: string
  description?: string
  status: 'pending' | 'running' | 'paused' | 'completed' | 'failed' | 'cancelled'
  is_test?: boolean
  seed_urls?: string[]
  max_depth?: number
  max_pages?: number
  max_concurrency?: number
  request_delay?: number
  timeout?: number
  created_at: string
  started_at?: string
  completed_at?: string
}

export interface CrawlBatchListResult {
  items: CrawlBatchSummary[]
  total: number
  page: number
  size: number
  pages: number
}

export interface CrawlBatchCreatePayload {
  project_id: string
  name: string
  seed_urls: string[]
  description?: string
  max_depth?: number
  max_pages?: number
  max_concurrency?: number
  request_delay?: number
  timeout?: number
  max_retries?: number
}

export interface CrawlBatchProgress {
  batch_id: string
  total_urls: number
  pending_urls: number
  completed_urls: number
  failed_urls: number
  progress_percentage?: number
}

export interface CrawlBatchItem {
  sequence: number
  url: string
  timestamp: string
  run_id: string
  data: Record<string, unknown>
}

class CrawlService extends BaseService {
  constructor() {
    // batches/metrics 端点挂在 /api/v1/crawl 前缀下
    super('/api/v1/crawl')
  }

  // 批次列表（支持项目筛选）。后端返回 PaginationResponse ——
  // BaseService.extractData 会剥掉最外层 `.data`，拿到的是
  // `{ items, pagination: { page, size, total, pages } }`（见
  // packages/antcode_core/.../schemas/common.py PaginationData）。
  // 前端上层只关心 items+total，这里把嵌套拍平，避免 BatchList 里出现
  // res.total ?? list.length 之类的 fallback（总数会永远等于当前页长度）。
  async listBatches(params?: {
    project_id?: string
    status?: string
    page?: number
    size?: number
  }): Promise<CrawlBatchListResult> {
    const raw = await this.get<{
      items: CrawlBatchSummary[]
      pagination: { page: number; size: number; total: number; pages: number }
    }>('/batches', { params })
    const items = raw?.items ?? []
    const pagination = raw?.pagination
    return {
      items,
      total: pagination?.total ?? items.length,
      page: pagination?.page ?? params?.page ?? 1,
      size: pagination?.size ?? params?.size ?? items.length,
      pages: pagination?.pages ?? 1,
    }
  }

  // 批次详情
  async getBatch(batchId: string): Promise<CrawlBatchSummary> {
    return this.get<CrawlBatchSummary>(`/batches/${batchId}`)
  }

  // 创建批次
  async createBatch(payload: CrawlBatchCreatePayload): Promise<CrawlBatchSummary> {
    return this.post<CrawlBatchSummary>('/batches', payload)
  }

  // 启动
  async startBatch(batchId: string): Promise<CrawlBatchSummary> {
    return this.post<CrawlBatchSummary>(`/batches/${batchId}/start`, {})
  }

  // 暂停
  async pauseBatch(batchId: string): Promise<CrawlBatchSummary> {
    return this.post<CrawlBatchSummary>(`/batches/${batchId}/pause`, {})
  }

  // 恢复
  async resumeBatch(batchId: string): Promise<CrawlBatchSummary> {
    return this.post<CrawlBatchSummary>(`/batches/${batchId}/resume`, {})
  }

  // 取消
  async cancelBatch(batchId: string): Promise<CrawlBatchSummary> {
    return this.post<CrawlBatchSummary>(`/batches/${batchId}/cancel`, {})
  }

  // 进度
  async getProgress(batchId: string): Promise<CrawlBatchProgress> {
    return this.get<CrawlBatchProgress>(`/batches/${batchId}/progress`)
  }

  // 批次维度聚合数据（R1-P2-28）
  async getBatchItems(
    batchId: string,
    limit = 100
  ): Promise<{ batch_id: string; items: CrawlBatchItem[]; count: number }> {
    return this.get<{ batch_id: string; items: CrawlBatchItem[]; count: number }>(
      `/batches/${batchId}/items`,
      { params: { limit } }
    )
  }

  // 导出：走浏览器下载（历史 URL 拼接方式）
  //
  // P2-15: 保留只为兼容旧调用点，新代码请用 exportBatch()。
  // window.open(exportBatchUrl(...)) 会绕过 axios 拦截器，Authorization
  // 头不会带上 → 服务端 401。
  exportBatchUrl(batchId: string, format: 'json' | 'csv'): string {
    return `/api/v1/crawl/batches/${batchId}/export?format=${format}`
  }

  /**
   * P2-15: 通过 axios 走 Bearer 认证下载导出文件，再本地触发下载。
   *
   * 原来 BatchList 直接 `window.open(exportBatchUrl(...))`，浏览器新标签页
   * 不会带 axios 拦截器注入的 Authorization 头，登录用户导出永远 401。
   * 这里改成 responseType: 'blob' 走鉴权拿字节流，再合成 <a download>
   * 触发下载。
   */
  async exportBatch(batchId: string, format: 'json' | 'csv'): Promise<void> {
    // BaseService.downloadFile 已经封装了 blob 请求 + content-disposition
    // 解析 + <a> 触发；这里直接复用，避免各页面再手写一遍下载胶水。
    await this.downloadFile(
      `/batches/${batchId}/export?format=${format}`,
      `crawl-batch-${batchId}.${format}`,
      { timeout: 0 }
    )
  }
}

export const crawlService = new CrawlService()
